src/utils.ts

```diff
@@ -12,38 +12,256 @@
+      )
+    }
+  }
+  return dtype ? tensor2d(data, undefined, dtype) : tensor2d(data)
+}
+
+export function convertToTensor1D_2D(
+  data: ScikitVecOrMatrix,
+  dtype?: DataType
+): Tensor1D | Tensor2D {
+  try {
+    const new1DTensor = convertToTensor1D(data as Tensor1D, dtype)
+    return new1DTensor
+  } catch (e) {
+    try {
+      const new2DTensor = convertToTensor2D(data as Tensor2D, dtype)
+      return new2DTensor
+    } catch (newE) {
+      throw new Error('ParamError: Can"t convert data into 1D or 2D tensor')
+    }
+  }
+}
+
+export function convertToNumericTensor1D_2D(
+  data: ScikitVecOrMatrix,
+  dtype?: DataType
+) {
+  const newTensor = convertToTensor1D_2D(data, dtype)
+  if (newTensor.dtype === 'string') {
+    throw new Error(
+      "ParamError: data has string dtype, can't convert to numeric Tensor"
+    )
+  }
+  return newTensor
+}
+
+export function assertSameShape(a: Tensor, b: Tensor) {
+  if (a.shape.length !== b.shape.length) {
+    throw new Error(
+      'ParamError: Shapes don"t match for these two Tensors. They are different dimensions'
+    )
+  }
+  for (let i = 0; i < a.shape.length; i++) {
+    if (a.shape[i] !== b.shape[i]) {
+      throw new Error('ParamError: Shapes do not match for these tensors')
+    }
+  }
+}
+
+export function assertSameType(a: Tensor, b: Tensor) {
+  if (a.dtype !== b.dtype) {
+    throw new Error('ParamError: These two Tensors do not have the same dtype')
+  }
+}
+
+export function assertSameShapeAndType(a: Tensor, b: Tensor) {
+  assertSameShape(a, b)
+  assertSameType(a, b)
+}
+
+export function convertToTensor(
+  data: TensorLike | Tensor | DataFrame | Series,
+  shape?: number[],
+  dtype?: keyof DataTypeMap
+): Tensor {
+  if (data instanceof DataFrame) {
+    return tensor2d(data.values, shape && [shape[0], shape[1]], dtype)
+  }
+  if (data instanceof Series) {
+    return tensor1d(data.values, dtype)
+  }
+  if (data instanceof Tensor) {
+    let newData = data
+    if (shape) {
+      newData = newData.reshape(shape)
+    }
+    if (dtype) {
+      newData = newData.asType(dtype)
     }
-}
\ No newline at end of file
+    return newData
+  }
+  return tensor(data, shape, dtype)
+}
+
+export function convertTensorToInputType(
+  tensor: Tensor,
+  inputData: ScikitVecOrMatrix
+) {
+  if (inputData instanceof Tensor) {
+    return tensor
+  } else if (inputData instanceof DataFrame) {
+    return new DataFrame(tensor, {
+      index: inputData.index,
+      columns: inputData.columns,
+    })
+  } else if (inputData instanceof Series) {
+    return new Series(tensor, {
+      index: inputData.index,
+    })
+  } else if (Array.isArray(inputData)) {
+    return tensor.arraySync()
+  } else {
+    return tensor
+  }
+}
+
+// For StandardScaler, and the others, the scikit-learn
+// api ignores Nan's for it's calculation, so we should too.
+
+export function minIgnoreNaN(tensor: Tensor, dim: number) {
```

Comment:
Is this not too repetitive? what if we have a function that removes all NaN values first before we apply a function on it. We currently do this in Danfo.js. [Here's](https://github.com/opensource9ja/danfojs/blob/4a4d45964b0ed381a3b3585eadf14ecae6504b62/danfojs-node/src/shared/utils.js#L642) the utility function we use, and [here's](https://github.com/opensource9ja/danfojs/blob/4a4d45964b0ed381a3b3585eadf14ecae6504b62/danfojs-node/src/core/frame.js#L837) an example of computing mean with NaN values removed. 

```javascript
function removeNaN(data) => removes all NaN values

//Then computations can be done directly:
tensor( removeNaN(data)).min()
tensor( removeNaN(data)).max()

```

Reply:
It was great talking to you as well. And yes go ahead and merge.